File: Algorithms/Sorting/QuickSort.go
package main
import (
	"fmt"	
)

func main(){
	arr:= []int{10,7,8,9,1,5} 
	n:= len(arr)
	quickSort(arr,0,n-1)
	fmt.Print("Sorted array: ")
	printArray(arr,n)
} 

func swap(a,b *int){
	t:= *a
	*a = *b
	*b = t
}

func partition(arr []int,low,high int) int{
	pivot:= arr[high]
	i:= (low - 1)
	j:= low

	for j <= high-1 {
		if arr[j] <= pivot {
			i++
			swap(&arr[i], &arr[j])
		}
		j++
	}
	swap(&arr[i+1], &arr[high])

	return (i+1)
}

func quickSort(arr []int,low,high int){
	if low < high {
		pi:= partition(arr,low,high)

		quickSort(arr,low,pi-1)
		quickSort(arr,pi+1,high)
	}
}

func printArray(arr []int,size int){
	i:=0

	for i < size {
		fmt.Print(arr[i]," ")
		i++
	}
}

File: Algorithms/QuickSort/quick_sort.cpp
#include <iostream>

using std::cout;

void quick_sort(int a[], int left, int right) {
    int i = left, j = right;
    int pivot = a[right];

    while (i <= j) {
        while (a[i] < pivot)
            i++; //carry on, this part already sorted
        while (a[j] > pivot)
            j--; //carry on, this part already sorted
        if (i <= j) { // a[i]>=pivot and a[j]<=pivot
        //swap a[i] with a[j]
            int tmp = a[i];
            a[i] = a[j];
            a[j] = tmp;
            i++;
            j--;
        }
    }

    if (left < j)
        quick_sort(a, left, j);
    if (i < right)
        quick_sort(a, i, right);
}

int main(void) {
    int a[10] = { 10, 7, 9, 4, 6, 2, 5, 6, 84, 12};
    int a_size = sizeof(a)/sizeof(a[0]);

    quick_sort(a, 0, a_size-1);

    for (int i = 0; i < a_size; i++) {
        cout << a[i] << " ";
    }
}
